docs/_docs/30-fasterkv-manual-locking.md

```diff
@@ -0,0 +1,179 @@
+---
+title: "FasterKV Manual Locking"
+permalink: /docs/fasterkv-manual-locking/
+excerpt: "FasterKV Manual Locking"
+last_modified_at: 2021-11-30
+toc: true
+---
+
+## Manual Record locking in FasterKV
+
+Manual locking in FasterKV refers to the user specifying when records will be locked. This is different from the per-operation locks that ensure consistency for concurrent operations, e.g. ConcurrentReader and ConcurrentWriter. Manual locks have a longer duration.
+
+Manual locking is done by obtaining the `ManualFasterOperations` instance from a `ClientSession`. This provides an implementation of `IFasterOperations` that:
+- Does not do automatic locking (except when updated records are inserted, as described below)
+- Does not do automatic epoch protection; instead, the user must call `UnsafeResumeThread` and `UnsafeSuspendThread`. In these, "Unsafe" refers to the fact it is the user's responsibility to make the correct calls.
+- Exposes `Lock()` and `Unlock()` APIs.
+
+Here are two use case examples:
+- Lock key1, key2, and key3, then Read key1 and key2 values, calculate the result, write them to key3, and unlock all keys. This ensures that key3 has a consistent value based on key1 and key2 values.
+- Lock key1, do a bunch of operations on other keys, then unlock key1. As long as the set of keys for this operation are partitioned by the choice for key1 and all updates on those keys are done only when the lock for key1 is held, this ensures the consistency of those keys' values.
+
+### Considerations
+
+All keys must be locked in a deterministic order, and unlocked in the reverse order, to avoid deadlocks.
+
+`ManualFasterOperations` inherits from `IDisposable`. All locks must be released and `UnsafeSuspendThread` must be called before `Dispose()` is called; `Dispose()` does *not* make these calls automatically.
+
+### Examples
+Here are examples of the above two use cases, taken from the unit tests in `ManualOperationsTests.cs`:
+
+Lock multiple keys:
+```cs
+    using (var manualOps = session.GetManualOperations())
+    {
+        manualOps.UnsafeResumeThread(out var epoch);
+
+        LockInfo lockInfo = default;
+        manualOps.Lock(24, LockType.Shared, retrieveData: true, ref lockInfo);
```

Comment:
With LockTable, do we really need `retrieveData`? Because we can assume false and write to the lock table, and later get the data if the operation wants it.